core/src/test/scala/unit/kafka/log/LogCleanerTest.scala

```diff
@@ -83,11 +84,19 @@ class LogCleanerTest {
       val numMetricsRegistered = LogCleaner.MetricNames.size
       verify(mockMetricsGroup, times(numMetricsRegistered)).newGauge(anyString(), any())
       
-      // verify that each metric is removed
+      // verify that each metric in `LogCleaner` is removed
       LogCleaner.MetricNames.foreach(verify(mockMetricsGroup).removeMetric(_))
 
+      // verify that each metric in `LogCleanerManager` is removed
+      val mockLogCleanerManagerMetricsGroup = mockMetricsGroupCtor.constructed.get(1)
+      LogCleanerManager.GaugeMetricNameNoTag.foreach(metricName => verify(mockLogCleanerManagerMetricsGroup).newGauge(ArgumentMatchers.eq(metricName), any()))
+      LogCleanerManager.GaugeMetricNameWithTag.keySet().asScala.foreach(metricName => verify(mockLogCleanerManagerMetricsGroup).newGauge(ArgumentMatchers.eq(metricName), any(), any()))
```

Comment:
does this work? In our case we will call this multiple times with same metric Name. The any() will match all such occurrences whereas we are saying that this invocation occurs only 1 time (default is times(1), when not specified)

Please help me understand why it's working.

Reply:
Sorry, why do you think we will call this multiple times with the same metric name? The way I read this code is that we will get the keys from GaugeMetricNameWithTag and GaugeMetricNameNoTag (which are unique because of the nature of a set and map) and for each one we will verify that it is called only once.

Reply:
GaugeMetricNameWithTag can contain multiple metrics with same name but different tags. Let's say `UncleanablePartitionsCountMetricName ` is the name and tag will be directory path. Over here, we are verifying that there is an exactly 1 call to newGauge() with `UncleanablePartitionsCountMetricName` as the metricname. But that is not correct because we are actually calling `newGauge` for this key name numDirs times. 

Hence, this test should ideally fail. What am I missing here?

Reply:
> GaugeMetricNameWithTag can contain multiple metrics with same name but different tags. Let's say `UncleanablePartitionsCountMetricName ` is the name and tag will be directory path. Over here, we are verifying that there is an exactly 1 call to newGauge() with `UncleanablePartitionsCountMetricName` as the metricname. But that is not correct because we are actually calling `newGauge` for this key name numDirs times.
> 
> Hence, this test should ideally fail. What am I missing here?

I think the reason is: "logDirs = Array(TestUtils.tempDir())", means that the `numDirs` size is 1. @divijvaidya

Reply:
ah ok. Can we please unit test this will multi dirs since the metric addition/removal here is dependent on number of directories and we want to ensure it works correctly for multiple dir case as well.

Reply:
I have updated this. @divijvaidya